src/mlpack/methods/ann/layer/not_adapted/parametric_relu_impl.hpp

```diff
@@ -21,66 +21,66 @@
 namespace mlpack {
 namespace ann /** Artificial Neural Network. */ {
 
-template<typename InputDataType, typename OutputDataType>
-PReLU<InputDataType, OutputDataType>::PReLU(
+template<typename InputType, typename OutputType>
+PReLUType<InputType, OutputType>::PReLUType(
     const double userAlpha) : userAlpha(userAlpha)
 {
   alpha.set_size(WeightSize(), 1);
   alpha(0) = userAlpha;
 }
 
-template<typename InputDataType, typename OutputDataType>
-void PReLU<InputDataType, OutputDataType>::Reset()
+template<typename InputType, typename OutputType>
+void PReLUType<InputType, OutputType>::SetWeights(
+    typename OutputType::elem_type* weightsPtr)
 {
+  alpha = arma::mat(weightsPtr, 1, 1, false, false);
+
   //! Set value of alpha to the one given by user.
-  alpha = arma::mat(alpha.memptr(), 1, 1, false, false);
+  // TODO: this doesn't even make any sense.  is it trainable or not?
+  // why is there userAlpha?  is that for initialization only?
+  alpha(0) = userAlpha;
 }
 
-template<typename InputDataType, typename OutputDataType>
 template<typename InputType, typename OutputType>
-void PReLU<InputDataType, OutputDataType>::Forward(
+void PReLUType<InputType, OutputType>::Forward(
     const InputType& input, OutputType& output)
 {
-  output = arma::max(input, alpha(0) * input);
+  // TODO: use transform()?
+  output = input;
+  arma::uvec negative = arma::find(input < 0);
+  output(negative) = input(negative) * alpha(0);
 }
 
-template<typename InputDataType, typename OutputDataType>
-template<typename DataType>
-void PReLU<InputDataType, OutputDataType>::Backward(
-    const DataType& input, const DataType& gy, DataType& g)
+template<typename InputType, typename OutputType>
+void PReLUType<InputType, OutputType>::Backward(
+    const InputType& input, const OutputType& gy, OutputType& g)
 {
-  DataType derivative;
+  OutputType derivative;
   derivative.set_size(arma::size(input));
   for (size_t i = 0; i < input.n_elem; ++i)
-  {
     derivative(i) = (input(i) >= 0) ? 1 : alpha(0);
-  }
 
   g = gy % derivative;
 }
 
-template<typename InputDataType, typename OutputDataType>
-template<typename eT>
-void PReLU<InputDataType, OutputDataType>::Gradient(
-    const arma::Mat<eT>& input,
-    const arma::Mat<eT>& error,
-    arma::Mat<eT>& gradient)
+template<typename InputType, typename OutputType>
+void PReLUType<InputType, OutputType>::Gradient(
+    const InputType& input,
+    const OutputType& error,
+    OutputType& gradient)
 {
-  if (gradient.n_elem == 0)
-  {
-    gradient = arma::zeros<arma::mat>(1, 1);
-  }
```

Comment:
Why is this removed?

Reply:
There's no need for it anymore---each layer can expect that `gradient` is sized correctly when `Gradient()` is called.  (Of course, I do need to document all these assumptions still!)